src/osd/PeeringState.h

```diff
@@ -54,6 +54,59 @@ struct PGPool {
   void update(CephContext *cct, OSDMapRef map);
 };
 
+// [primary only] content recovery state
+struct BufferedRecoveryMessages {
+  map<int, map<spg_t, pg_query_t> > query_map;
+  map<int, vector<pair<pg_notify_t, PastIntervals> > > info_map;
+  map<int, vector<pair<pg_notify_t, PastIntervals> > > notify_list;
+};
+
+struct PeeringCtx {
+  map<int, map<spg_t, pg_query_t> > query_map;
+  map<int, vector<pair<pg_notify_t, PastIntervals> > > info_map;
+  map<int, vector<pair<pg_notify_t, PastIntervals> > > notify_list;
+  ObjectStore::Transaction transaction;
+  HBHandle* handle = nullptr;
+
+  PeeringCtx() = default;
+
+  void reset_transaction() {
+    transaction = ObjectStore::Transaction();
+  }
+
+  void accept_buffered_messages(BufferedRecoveryMessages &m) {
+    for (map<int, map<spg_t, pg_query_t> >::iterator i = m.query_map.begin();
+	 i != m.query_map.end();
+	 ++i) {
+      map<spg_t, pg_query_t> &omap = query_map[i->first];
+      for (map<spg_t, pg_query_t>::iterator j = i->second.begin();
+	   j != i->second.end();
+	   ++j) {
+	omap[j->first] = j->second;
+      }
+    }
+    for (map<int, vector<pair<pg_notify_t, PastIntervals> > >::iterator i
+	   = m.info_map.begin();
+	 i != m.info_map.end();
+	 ++i) {
+      vector<pair<pg_notify_t, PastIntervals> > &ovec =
+	info_map[i->first];
+      ovec.reserve(ovec.size() + i->second.size());
+      ovec.insert(ovec.end(), i->second.begin(), i->second.end());
+    }
+    for (map<int, vector<pair<pg_notify_t, PastIntervals> > >::iterator i
```

Comment:
ditto.

Reply:
Yeah, that actually made it far clearer.